File: problems/2-medium/143-Reorder-List/solution.java
public class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public void reorderList(ListNode head) {
        
        ListNode mid = findMidNode(head);
        ListNode reversed_list = reverseList(mid);
        mergeTwoLists(head,reversed_list,mid);

    }

    private ListNode findMidNode(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
           
        return slow;
    }

    private ListNode mergeTwoLists(ListNode list1, ListNode list2, ListNode mid) {
    
        ListNode dummy = new ListNode();
        ListNode current = dummy;

        while (list2 != null && list1 != mid) {
            current.next = list1;
            current = current.next;
            list1 = list1.next;

            current.next = list2;
            current = current.next;
            list2 = list2.next;
        }

        if (list2 != null) {
            current.next = list2;
        }
        
        return dummy.next;

    }

    private ListNode reverseList(ListNode head) {

        ListNode prev = null;
        ListNode current = head;
        ListNode nxt = null;

        while (current != null) {
            nxt = current.next;
            current.next = prev;
            prev = current;
            current = nxt;
        }
        
        return prev;

    }



}

File: problems/2-medium/143-Reorder-List/solution.cpp
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode* getMidNode(ListNode* head);
ListNode* reverseList(ListNode* head);

void reorderList(ListNode* head) {
    // get middle node
    ListNode* mid = getMidNode(head);
    // reverse second half
    ListNode* reversed = reverseList(mid);
    
    // interleave nodes from head and reversed 
    ListNode dummy_head = ListNode(0, nullptr);
    ListNode* dummy_ptr = &dummy_head;
    // (note: reversed second half >= first half)
    while (head != mid) {
        // add node from first half
        dummy_ptr->next = head;
        dummy_ptr = dummy_ptr->next;
        head = head->next;

        // add node from reversed second half
        dummy_ptr->next = reversed;
        dummy_ptr = dummy_ptr->next;
        reversed = reversed->next;
    }
    if (reversed) dummy_ptr->next = reversed;
}

ListNode* getMidNode(ListNode* head)  {
    // find mid point
    ListNode* slow = head;
    ListNode* fast = head;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow;
}

ListNode* reverseList(ListNode* head) {
    // reverse list
    ListNode* curr = head;
    ListNode* prev = nullptr;
    while (curr) {
        ListNode* next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    return prev;
}